### PR TITLE
Implement more features in Brian2

### PR DESCRIPTION
**Pull request by wilkieolin**
Changed equations and methods in brian2 backend to implement:
*Reset by subtraction
*Non-Poisson (constant current injection) input layer
*Neuron biases

### CODE REVIEW

File: snntoolbox/simulation/target_simulators/brian2_target_sim.py [hunk 1]
```diff
@@ -378,14 +388,11 @@ def set_spiketrain_stats_input(self):
...
-    """
+    def set_biases(self, biases):
+        """Set biases.
+        """
 
-    if any(biases):  # TODO: Implement biases.
-        warnings.warn("Biases not implemented.", RuntimeWarning)
+        if any(biases):
+            if self.layers[-1].current.shape != biases.shape:
+                raise AttributeError
```
Comment (rbodo) [inline snntoolbox/simulation/target_simulators/brian2_target_sim.py]: Why is this an ``AttributeError``? Consider replacing with ``assert``.

File: snntoolbox/simulation/target_simulators/brian2_target_sim.py [hunk 2]
```diff
@@ -378,14 +388,11 @@ def set_spiketrain_stats_input(self):
...
-
-    This has not been tested yet.
-    """
+    def set_biases(self, biases):
+        """Set biases.
+        """
 
-    if any(biases):  # TODO: Implement biases.
-        warnings.warn("Biases not implemented.", RuntimeWarning)
+        if any(biases):
```
[the remark this answers is not in the document]
Comment (wilkieolin) [reply snntoolbox/simulation/target_simulators/brian2_target_sim.py]: I'm also seeing that there is reduced accuracy using the new methods. I'm trying to track down which of the changes is causing this right now, will update when I find it.

File: snntoolbox/simulation/target_simulators/brian2_target_sim.py [hunk 3]
```diff
@@ -98,7 +109,7 @@ def add_layer(self, layer):
             return
 
         self.layers.append(self.sim.NeuronGroup(
-            np.prod(layer.output_shape[1:]), model=self.eqs, method='linear',
```
[the remark this answers is not in the document]
Comment (wilkieolin) [reply snntoolbox/simulation/target_simulators/brian2_target_sim.py]: I had thought there might be a speedup for a solver that can conditionally be linear, but didn't see this in testing. Switched to default 'euler' solver.

File: snntoolbox/simulation/target_simulators/brian2_target_sim.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wilkieolin) [reply snntoolbox/simulation/target_simulators/brian2_target_sim.py]: I've done more testing and it looks like two issues were involved in the accuracy drop. One was the permutation of weights in build_dense which is done after a flatten layer. I found that including it caused the network's performance to drop to random chance (92% to 11%, using Keras 2.3.1). 

Second, the default value of `dt = 1 ms` isn't sufficient for the Brian2 solver to integrate the potentials accurately when reset by subtraction and biases are being used. 

Test results, Convolutional network included in example
`ANN accuracy 99.61%`
`SNN n_test = 256`

| dt value (ms) | Reset | Bias | Accuracy (%) |
| ----- | ----- | ---- | ---- |
| 1 | To Zero | No Bias | 92 |
| 1 | Subtraction | No Bias | 84 | 
| 0.1 | Subtraction | No Bias | 98.8 |
| 0.1 | Subtraction | Bias | 99.2 |

Further testing shows that `dt` of 0.1 ms or lower seems best.

| dt value (ms) | Reset | Bias | Accuracy (%) |
| ----- | ----- | ---- | ---- |
| 1 | Subtraction | Bias | 83.98 |
| 0.5 | Subtraction | Bias | 92.10 | 
| 0.1 | Subtraction | Bias | 99.22 |

[the remark this answers is not in the document]
Comment (rbodo) [reply snntoolbox/simulation/target_simulators/brian2_target_sim.py]: Thanks for this analysis. With ``dt = 0.1`` and the correct time scale for the biases, I also get 99% accuracy after 200 samples. However, I need to keep the "flatten-fix" in place. Without it, accuracy drops to chance level. 

I've done a clean install of SNN Toolbox (from pip or github) in a virtual environment with the most recent versions of the dependencies. Here's the output of my conda list:

```
absl-py                   0.8.1                    pypi_0    pypi
astor                     0.8.0                    pypi_0    pypi
atomicwrites              1.3.0                    pypi_0    pypi
attrs                     19.3.0                   pypi_0    pypi
brian2                    2.2.2.1                  pypi_0    pypi
ca-certificates           2019.10.16                    0
certifi                   2019.9.11                py37_0
colorama                  0.4.1                    pypi_0    pypi
cycler                    0.10.0                   pypi_0    pypi
cython                    0.29.13                  pypi_0    pypi
filelock                  3.0.12                   pypi_0    pypi
future                    0.18.1                   pypi_0    pypi
gast                      0.2.2                    pypi_0    pypi
google-pasta              0.1.7                    pypi_0    pypi
grpcio                    1.24.3                   pypi_0    pypi
h5py                      2.10.0                   pypi_0    pypi
importlib-metadata        0.23                     pypi_0    pypi
jinja2                    2.10.3                   pypi_0    pypi
keras                     2.3.1                    pypi_0    pypi
keras-applications        1.0.8                    pypi_0    pypi
keras-preprocessing       1.1.0                    pypi_0    pypi
kiwisolver                1.1.0                    pypi_0    pypi
markdown                  3.1.1                    pypi_0    pypi
markupsafe                1.1.1                    pypi_0    pypi
matplotlib                3.1.1                    pypi_0    pypi
more-itertools            7.2.0                    pypi_0    pypi
mpmath                    1.1.0                    pypi_0    pypi
numpy                     1.17.3                   pypi_0    pypi
openssl                   1.1.1d               he774522_3
opt-einsum                3.1.0                    pypi_0    pypi
packaging                 19.2                     pypi_0    pypi
pillow                    6.2.1                    pypi_0    pypi
pip                       19.3.1                   py37_0
pluggy                    0.13.0                   pypi_0    pypi
protobuf                  3.10.0                   pypi_0    pypi
py                        1.8.0                    pypi_0    pypi
py-cpuinfo                5.0.0                    pypi_0    pypi
pyparsing                 2.4.2                    pypi_0    pypi
pytest                    5.2.2                    pypi_0    pypi
python                    3.7.4                h5263a28_0
python-dateutil           2.8.0                    pypi_0    pypi
pyyaml                    5.1.2                    pypi_0    pypi
scipy                     1.3.1                    pypi_0    pypi
setuptools                41.4.0                   py37_0
six                       1.12.0                   pypi_0    pypi
snntoolbox                0.3.1                    pypi_0    pypi
sqlite                    3.30.0               he774522_0
sympy                     1.4                      pypi_0    pypi
tensorboard               2.0.0                    pypi_0    pypi
tensorflow                2.0.0                    pypi_0    pypi
tensorflow-estimator      2.0.1                    pypi_0    pypi
termcolor                 1.1.0                    pypi_0    pypi
toml                      0.10.0                   pypi_0    pypi
tox                       3.14.0                   pypi_0    pypi
vc                        14.1                 h0510ff6_4
virtualenv                16.7.7                   pypi_0    pypi
vs2015_runtime            14.16.27012          hf0eaf9b_0
wcwidth                   0.1.7                    pypi_0    pypi
werkzeug                  0.16.0                   pypi_0    pypi
wheel                     0.33.6                   py37_0
wincertstore              0.2                      py37_0
wrapt                     1.11.2                   pypi_0    pypi
zipp                      0.6.0                    pypi_0    pypi
```

I'm limited to using a CPU on Windows at the moment. What platform are you on? Can you likewise do a clean install in a new virtual environment, and set the CPU flag in your keras config? I'm curious where this discrepancy comes from.

[the remark this answers is not in the document]
Comment (wilkieolin) [reply snntoolbox/simulation/target_simulators/brian2_target_sim.py]: I re-ran an example with a fresh environment and also found that the swap is now required to get good performance, so I've restored the piece of code to do the swap in this branch. 

Both my environments are using CPU-only Tensorflow with Python 3.7 on Ubuntu 18.04. I've attached the environments I'm using (new and old) for comparison. I haven't confirmed this, but I suspect that the different might be due to the Tensorflow backend being different (2.0 vs 1.14.0). I suspect this probably won't be an issue for most new users doing installs with TF 2.0. 

**New Environment (Needs Flatten Swap)**
```
# packages in environment at /home/wolin/anaconda3/envs/snntb:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
absl-py                   0.8.1                    pypi_0    pypi
astor                     0.8.0                    pypi_0    pypi
attrs                     19.3.0                   pypi_0    pypi
backcall                  0.1.0                    pypi_0    pypi
binutils_impl_linux-64    2.31.1               h7fc9f1b_2    conda-forge
binutils_linux-64         2.31.1              h6176602_12    conda-forge
bleach                    3.1.0                    pypi_0    pypi
brian2                    2.2.2.1          py37hc9558a2_1    conda-forge
ca-certificates           2019.9.11            hecc5488_0    conda-forge
cachetools                3.1.1                    pypi_0    pypi
certifi                   2019.9.11                py37_0    conda-forge
chardet                   3.0.4                    pypi_0    pypi
cycler                    0.10.0                   pypi_0    pypi
cython                    0.29.13          py37he1b5a44_0    conda-forge
decorator                 4.4.1                    pypi_0    pypi
defusedxml                0.6.0                    pypi_0    pypi
entrypoints               0.3                      pypi_0    pypi
fastcache                 1.1.0            py37h516909a_0    conda-forge
future                    0.18.2                   pypi_0    pypi
gast                      0.2.2                    pypi_0    pypi
gcc_impl_linux-64         7.3.0                habb00fd_2    conda-forge
gcc_linux-64              7.3.0               h553295d_12    conda-forge
gmp                       6.1.2             hf484d3e_1000    conda-forge
gmpy2                     2.1.0b1          py37h04dde30_0    conda-forge
google-auth               1.6.3                    pypi_0    pypi
google-auth-oauthlib      0.4.1                    pypi_0    pypi
google-pasta              0.1.7                    pypi_0    pypi
grpcio                    1.24.3                   pypi_0    pypi
gsl                       2.6                  h294904e_0    conda-forge
gxx_impl_linux-64         7.3.0                hdf63c60_2    conda-forge
gxx_linux-64              7.3.0               h553295d_12    conda-forge
h5py                      2.10.0                   pypi_0    pypi
idna                      2.8                      pypi_0    pypi
importlib-metadata        0.23                     pypi_0    pypi
ipykernel                 5.1.3                    pypi_0    pypi
ipython                   7.9.0                    pypi_0    pypi
ipython-genutils          0.2.0                    pypi_0    pypi
ipywidgets                7.5.1                    pypi_0    pypi
jedi                      0.15.1                   pypi_0    pypi
jinja2                    2.10.3                     py_0    conda-forge
jsonschema                3.1.1                    pypi_0    pypi
jupyter                   1.0.0                    pypi_0    pypi
jupyter-client            5.3.4                    pypi_0    pypi
jupyter-console           6.0.0                    pypi_0    pypi
jupyter-core              4.6.1                    pypi_0    pypi
keras                     2.3.1                    pypi_0    pypi
keras-applications        1.0.8                    pypi_0    pypi
keras-preprocessing       1.1.0                    pypi_0    pypi
kiwisolver                1.1.0                    pypi_0    pypi
libblas                   3.8.0               14_openblas    conda-forge
libcblas                  3.8.0               14_openblas    conda-forge
libedit                   3.1.20181209         hc058e9b_0  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 9.1.0                hdf63c60_0  
libgfortran-ng            7.3.0                hdf63c60_2    conda-forge
liblapack                 3.8.0               14_openblas    conda-forge
libopenblas               0.3.7                h6e990d7_2    conda-forge
libstdcxx-ng              9.1.0                hdf63c60_0  
markdown                  3.1.1                    pypi_0    pypi
markupsafe                1.1.1            py37h14c3975_0    conda-forge
matplotlib                3.1.1                    pypi_0    pypi
mistune                   0.8.4                    pypi_0    pypi
more-itertools            7.2.0                    pypi_0    pypi
mpc                       1.1.0             h04dde30_1006    conda-forge
mpfr                      4.0.2                he80fd80_0    conda-forge
mpmath                    1.1.0                      py_0    conda-forge
nbconvert                 5.6.1                    pypi_0    pypi
nbformat                  4.4.0                    pypi_0    pypi
ncurses                   6.1                  he6710b0_1  
notebook                  6.0.1                    pypi_0    pypi
numpy                     1.17.3           py37h95a1406_0    conda-forge
oauthlib                  3.1.0                    pypi_0    pypi
openssl                   1.1.1d               h7b6447c_3  
opt-einsum                3.1.0                    pypi_0    pypi
pandocfilters             1.4.2                    pypi_0    pypi
parso                     0.5.1                    pypi_0    pypi
pexpect                   4.7.0                    pypi_0    pypi
pickleshare               0.7.5                    pypi_0    pypi
pip                       19.3.1                   py37_0  
prometheus-client         0.7.1                    pypi_0    pypi
prompt-toolkit            2.0.10                   pypi_0    pypi
protobuf                  3.10.0                   pypi_0    pypi
ptyprocess                0.6.0                    pypi_0    pypi
pyasn1                    0.4.7                    pypi_0    pypi
pyasn1-modules            0.2.7                    pypi_0    pypi
pygments                  2.4.2                    pypi_0    pypi
pyparsing                 2.4.2                      py_0    conda-forge
pyrsistent                0.15.5                   pypi_0    pypi
python                    3.7.5                h0371630_0  
python-dateutil           2.8.0                    pypi_0    pypi
pyyaml                    5.1.2                    pypi_0    pypi
pyzmq                     18.1.0                   pypi_0    pypi
qtconsole                 4.5.5                    pypi_0    pypi
readline                  7.0                  h7b6447c_5  
requests                  2.22.0                   pypi_0    pypi
requests-oauthlib         1.2.0                    pypi_0    pypi
rsa                       4.0                      pypi_0    pypi
scipy                     1.3.1            py37h921218d_2    conda-forge
send2trash                1.5.0                    pypi_0    pypi
setuptools                41.6.0                   py37_0  
six                       1.12.0                   pypi_0    pypi
sqlite                    3.30.1               h7b6447c_0  
sympy                     1.4                      py37_0    conda-forge
tensorboard               2.0.1                    pypi_0    pypi
tensorflow                2.0.0                    pypi_0    pypi
tensorflow-estimator      2.0.1                    pypi_0    pypi
termcolor                 1.1.0                    pypi_0    pypi
terminado                 0.8.2                    pypi_0    pypi
testpath                  0.4.2                    pypi_0    pypi
tk                        8.6.8                hbc83047_0  
tornado                   6.0.3                    pypi_0    pypi
traitlets                 4.3.3                    pypi_0    pypi
urllib3                   1.25.6                   pypi_0    pypi
wcwidth                   0.1.7                    pypi_0    pypi
webencodings              0.5.1                    pypi_0    pypi
werkzeug                  0.16.0                   pypi_0    pypi
wheel                     0.33.6                   py37_0  
widgetsnbextension        3.5.1                    pypi_0    pypi
wrapt                     1.11.2                   pypi_0    pypi
xz                        5.2.4                h14c3975_4  
zipp                      0.6.0                    pypi_0    pypi
zlib                      1.2.11               h7b6447c_3  
```
**Old Environment (Doesn't need flatten swap)**
```
# packages in environment at /home/wolin/anaconda3/envs/tf2:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
absl-py                   0.8.0                    pypi_0    pypi
astor                     0.8.0                    pypi_0    pypi
attrs                     19.1.0                   pypi_0    pypi
backcall                  0.1.0                    pypi_0    pypi
binutils_impl_linux-64    2.31.1               h6176602_1  
binutils_linux-64         2.31.1               h6176602_8  
blas                      1.0                         mkl  
bleach                    3.1.0                    pypi_0    pypi
brian2                    2.2.2.1          py37hc9558a2_1    conda-forge
ca-certificates           2019.9.11            hecc5488_0    conda-forge
certifi                   2019.9.11                py37_0    conda-forge
connplotter               0.7a                     pypi_0    pypi
cycler                    0.10.0                   py37_0  
cython                    0.29.13          py37he6710b0_0  
dbus                      1.13.6               h746ee38_0  
decorator                 4.4.0                    pypi_0    pypi
defusedxml                0.6.0                    pypi_0    pypi
dill                      0.3.1.1                  pypi_0    pypi
entrypoints               0.3                      pypi_0    pypi
expat                     2.2.6                he6710b0_0  
fastcache                 1.1.0            py37h7b6447c_0  
fontconfig                2.13.0               h9420a91_0  
freetype                  2.9.1                h8a8886c_1  
future                    0.17.1                   pypi_0    pypi
gast                      0.3.2                    pypi_0    pypi
gcc_impl_linux-64         7.3.0                habb00fd_1  
gcc_linux-64              7.3.0                h553295d_8  
glib                      2.56.2               hd408876_0  
gmp                       6.1.2                h6c8ec71_1  
gmpy2                     2.0.8            py37h10f8cd9_2  
google-pasta              0.1.7                    pypi_0    pypi
grpcio                    1.24.0                   pypi_0    pypi
gsl                       2.4                  h14c3975_4  
gst-plugins-base          1.14.0               hbbd80ab_1  
gstreamer                 1.14.0               hb453b48_1  
gxx_impl_linux-64         7.3.0                hdf63c60_1  
gxx_linux-64              7.3.0                h553295d_8  
h5py                      2.10.0                   pypi_0    pypi
icu                       58.2                 h9c2bf20_1  
intel-openmp              2019.4                      243  
ipykernel                 5.1.2                    pypi_0    pypi
ipython                   7.8.0                    pypi_0    pypi
ipython-genutils          0.2.0                    pypi_0    pypi
ipywidgets                7.5.1                    pypi_0    pypi
jedi                      0.15.1                   pypi_0    pypi
jinja2                    2.10.3                     py_0  
joblib                    0.14.0                   pypi_0    pypi
jpeg                      9b                   h024ee3a_2  
jsonschema                3.0.2                    pypi_0    pypi
jupyter                   1.0.0                    pypi_0    pypi
jupyter-client            5.3.3                    pypi_0    pypi
jupyter-console           6.0.0                    pypi_0    pypi
jupyter-core              4.5.0                    pypi_0    pypi
keras                     2.3.1                    pypi_0    pypi
keras-applications        1.0.8                    pypi_0    pypi
keras-preprocessing       1.1.0                    pypi_0    pypi
kiwisolver                1.1.0            py37he6710b0_0  
krb5                      1.16.3            h05b26f9_1001    conda-forge
lazyarray                 0.3.3                    pypi_0    pypi
libedit                   3.1.20181209         hc058e9b_0  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 9.1.0                hdf63c60_0  
libgfortran-ng            7.3.0                hdf63c60_0  
libpng                    1.6.37               hbc83047_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
libtool                   2.4.6             h14c3975_1002    conda-forge
libuuid                   1.0.3                h1bed415_2  
libxcb                    1.13                 h1bed415_1  
libxml2                   2.9.9                hea5a465_1  
markdown                  3.1.1                    pypi_0    pypi
markupsafe                1.1.1            py37h7b6447c_0  
matplotlib                3.1.1            py37h5429711_0  
mistune                   0.8.4                    pypi_0    pypi
mkl                       2019.4                      243  
mkl-service               2.3.0            py37he904b0f_0  
mkl_fft                   1.0.14           py37ha843d7b_0  
mkl_random                1.1.0            py37hd6b4f25_0  
mock                      3.0.5                    pypi_0    pypi
mpc                       1.1.0                h10f8cd9_1  
mpfr                      4.0.1                hdf1c602_3  
mpi                       1.0                     openmpi    conda-forge
mpi4py                    3.0.1            py37hd955b32_0    conda-forge
mpmath                    1.1.0                    py37_0  
nbconvert                 5.6.0                    pypi_0    pypi
nbformat                  4.4.0                    pypi_0    pypi
ncurses                   6.1                  he6710b0_1  
neo                       0.8.0                    pypi_0    pypi
nest-simulator            2.16.0          mpi_openmpi_py37hea85dc9_1000    conda-forge
notebook                  6.0.1                    pypi_0    pypi
numpy                     1.17.2           py37haad9e8e_0  
numpy-base                1.17.2           py37hde5b4d6_0  
openmpi                   3.1.4                hc99cbb1_0    conda-forge
openssh                   7.9p1                h0fa992c_1    conda-forge
openssl                   1.1.1c               h516909a_0    conda-forge
pandas                    0.25.1           py37hb3f55d8_0    conda-forge
pandocfilters             1.4.2                    pypi_0    pypi
parso                     0.5.1                    pypi_0    pypi
patsy                     0.5.1                      py_0    conda-forge
pcre                      8.43                 he6710b0_0  
pexpect                   4.7.0                    pypi_0    pypi
pickleshare               0.7.5                    pypi_0    pypi
pip                       19.2.3                   py37_0  
prometheus-client         0.7.1                    pypi_0    pypi
prompt-toolkit            2.0.9                    pypi_0    pypi
protobuf                  3.9.2                    pypi_0    pypi
ptyprocess                0.6.0                    pypi_0    pypi
pygments                  2.4.2                    pypi_0    pypi
pynn                      0.9.4                    pypi_0    pypi
pyparsing                 2.4.2                      py_0  
pyqt                      5.9.2            py37h05f1152_2  
pyrsistent                0.15.4                   pypi_0    pypi
python                    3.7.4                h265db76_1  
python-dateutil           2.8.0                    py37_0  
pytz                      2019.3                     py_0  
pyyaml                    5.1.2                    pypi_0    pypi
pyzmq                     18.1.0                   pypi_0    pypi
qt                        5.9.7                h5867ecd_1  
qtconsole                 4.5.5                    pypi_0    pypi
quantities                0.12.3                   pypi_0    pypi
readline                  7.0                  h7b6447c_5  
scikit-learn              0.21.3                   pypi_0    pypi
scipy                     1.3.1            py37h7c811a0_0  
send2trash                1.5.0                    pypi_0    pypi
setuptools                41.2.0                   pypi_0    pypi
sip                       4.19.8           py37hf484d3e_0  
six                       1.12.0                   pypi_0    pypi
sklearn                   0.0                      pypi_0    pypi
sqlite                    3.30.0               h7b6447c_0  
statsmodels               0.10.1           py37hc1659b7_0    conda-forge
sympy                     1.4                      py37_0  
tensorboard               1.14.0                   pypi_0    pypi
tensorflow                1.14.0                   pypi_0    pypi
tensorflow-estimator      1.14.0                   pypi_0    pypi
termcolor                 1.1.0                    pypi_0    pypi
terminado                 0.8.2                    pypi_0    pypi
testpath                  0.4.2                    pypi_0    pypi
tk                        8.6.9             hed695b0_1003    conda-forge
topology                  2.16.0                   pypi_0    pypi
tornado                   6.0.3            py37h7b6447c_0  
traitlets                 4.3.2                    pypi_0    pypi
wcwidth                   0.1.7                    pypi_0    pypi
webencodings              0.5.1                    pypi_0    pypi
werkzeug                  0.16.0                   pypi_0    pypi
wheel                     0.33.6                   pypi_0    pypi
widgetsnbextension        3.5.1                    pypi_0    pypi
wrapt                     1.11.2                   pypi_0    pypi
xz                        5.2.4                h14c3975_4  
zlib                      1.2.11               h7b6447c_3  
```

**Pull request merged**